R/helpers.R

```diff
@@ -196,3 +196,248 @@ list_to_hive_partition <- function(x) {
+        destPath = file.path(!!table_folder, "_table_changes", hivePartitions),
+        filePath = file.path(destPath, file)
+      )
+  } else {
+    files <- files %>%
+      dplyr::mutate(
+        files, file = paste0(id, ".parquet"),
+        destPath = file.path(!!table_folder, "_table", hivePartitions),
+        filePath = file.path(destPath, file)
+      )
+  }
+
+  # both paths are generated to check if files can be skipped from
+  # download step later on, there are two types of skipping:
+  # 1. file exists in current directory (_table_changes or root)
+  # 2. can be copied between _table_changes and root directory
+  existing_files <- list.files(path = table_folder, recursive = TRUE) %>%
+    data.frame(file = .) %>%
+    dplyr::mutate(
+      id = gsub(".*?(.*\\/)?([a-z0-9]*?)(_.*?)?(\\.parquet)", "\\2", file, perl = TRUE),
+      is_in_cdf = grepl("^_table_changes", file)
+    )
+
+  existing_ids_root <- dplyr::filter(existing_files, !is_in_cdf)$id
+  existing_ids_cdf <- dplyr::filter(existing_files, is_in_cdf)$id
+
+  files <- files %>%
+    dplyr::mutate(
+      relativePath = sub("^\\/", "", file.path(hivePartitions, file)),
+      alreadyExistsRoot = id %in% existing_ids_root,
+      alreadyExistsInCDF = id %in% existing_ids_cdf,
+      alreadyExists = alreadyExistsRoot | alreadyExistsInCDF,
+      to_download = !alreadyExists,
+      to_copy_to_cdf = !alreadyExistsInCDF & alreadyExistsRoot,
+      to_copt_to_root = !alreadyExistsRoot & alreadyExistsInCDF
+    )
+
+  files
+
+}
+
+# TODO: Documentation
+download_new_files <- function(new_query_files, changes) {
+
+  # create progress bar
+  pb <- progress::progress_bar$new(
+    format = "  downloading files (:elapsedfull) [:bar] :current/:total (:percent) [:eta]",
+    total = nrow(new_query_files),
+    clear = FALSE,
+    width = 100
+  )
+  pb$tick(0)
+
+  # download new files
+  new_query_files %>%
+    purrr::pwalk(function(url, filePath, ...) {
+      download.file(url, destfile = filePath, quiet = TRUE)
+      pb$tick()
+    })
+
+}
+
+# TODO: Documentation
+version_is_valid <- function(x) {
+  floor(x) == x
+}
+
+# TODO: Documentation
+timestamp_is_valid <- function(x) {
+  # it should be defined as a character
+  if (is.character(x)) {
+    x <- as.POSIXct(x, format = "%Y-%m-%d %H:%M:%S")
+    return(!is.na(x))
+  }
+  FALSE
+}
+
+# TODO: Documentation
+validate_cdf_options <- function(starting_version, ending_version,
+                                 starting_timestamp, ending_timestamp) {
+
+  # minimum requirement is to have starting_* parameter
+  # can only specify either version or timestamp, not both
+  if (!xor(is.null(starting_version), is.null(starting_timestamp))) {
+    rlang::abort(
+      paste(
+        "INVALID_PARAMETER_VALUE: Must pass either starting_version or",
+        "starting_timestamp using set_cdf_options() before performing CDF read."
+      )
+    )
+  }
+
+  if (all(!is.null(c(ending_version, ending_timestamp)))) {
```

Comment:
`is.null` only returns one logical value (rather than a vector of equal length), so something like `ds_tbl$set_cdf_options(starting_version = 8, ending_version = 9)` now returns an error. Probably need to throw an `*apply` in there.